code.py

```diff
@@ -0,0 +1,182 @@
+import os
+import flask
+import hashlib
+import requests
+
+app = flask.Flask(__name__)
+
+def clean_thing(thing):
+    # CLEAN
+    return thing
+
+def do_thing(thing):
+    pass
+
+@app.route("/route_param/<route_param>")
+def route_param(route_param):
+    print("blah")
+    # ruleid: tainted-url-host
+    cleaned = clean_thing(route_param)
+    url = "https://%s/path" % cleaned
+    requests.get(url)
+
+    # ruleid: tainted-url-host
+    url = "http://%r/path" % cleaned
+    requests.get(url)
+
+    return True
+
+@app.route("/route_param_ok/<route_param>")
+def route_param_ok(route_param):
+    print("blah")
+    # ok: tainted-url-host
+    return "<a href='https://example.com'>Click me!</a>"
+
+@app.route("/route_param_format/<route_param>")
+def route_param_format(route_param):
+    print("blah")
+    # ruleid: tainted-url-host
+    do_thing("https://%s/path" % route_param)
+    return "<a href='https://{}/path'>Click me!</a>".format(route_param)
+
+@app.route("/route_param_format_ok_in_path/<route_param>")
+def route_param_format_ok_in_path(route_param):
+    print("blah")
+    # ok: tainted-url-host
+    return "<a href='https://example.com/{}/path'>Click me!</a>".format(route_param)
+
+@app.route("/route_param_percent_format/<route_param>")
+def route_param_percent_format(route_param):
+    print("blah")
+    cleaned = clean_thing(route_param)
+    # ruleid: tainted-url-host
+    do_thing("https://%s/path" % route_param)
+    do_thing("https://%s/path" % cleaned)
+    return "<a href='https://%s/path'>Click me!</a>" % cleaned
+
+@app.route("/route_param_percent_format_ok_in_path/<route_param>")
+def route_param_percent_format_ok_in_path(route_param):
+    print("blah")
+    # ok: tainted-url-host
+    return "<a href='https://example.com/%s/path'>Click me!</a>" % route_param
+
+@app.route("/get_param_inline", methods=["GET"])
+def get_param_inline():
+    # ruleid: tainted-url-host
+    return "<a href='https://%s/path'>Click me!</a>" % flask.request.args.get("param")
+
+@app.route("/get_param_inline_concat", methods=["GET"])
+def get_param_inline_concat():
+    # ruleid: tainted-url-host
+    return "<a href='http://" + flask.request.args.get("param") + "'>Click me!</a>"
+
+@app.route("/get_param_inline_concat_ok_in_path", methods=["GET"])
+def get_param_inline_concat_ok_in_path():
+    # ok: tainted-url-host
+    return "<a href='http://example.com/" + flask.request.args.get("param") + "'>Click me!</a>"
+
+@app.route("/get_param_template", methods=["GET"])
+def get_param_template():
+    # ruleid: tainted-url-host
+    return f"<a href='https://{flask.request.args.get('param')}/path'>Click me!</a>"
+
+@app.route("/get_param_template_ok_in_path", methods=["GET"])
+def get_param_template_ok_in_path():
+    # ok: tainted-url-host
+    return f"<a href='https://example.com/{flask.request.args.get('param')}/path'>Click me!</a>"
+
+@app.route("/get_param_concat", methods=["GET"])
```

Comment:
Example dataflow finding [Lines X-X]
<details>

<summary>View Dataflow Graph</summary>

```mermaid
flowchart LR
    classDef invis fill:white, stroke: none
    classDef default fill:#e7f5ff, color:#1c7fd6, stroke: none

    subgraph File0["<b>foo.py</b>"]
        direction LR
        %% Source

        subgraph Source
            direction LR

            v0("<b>[Line: 26]</b> danger")
        end
        %% Intermediate

        subgraph Traces0[Traces]
            direction TB

            v2("<b>[Line: 26]</b> d2")
        end
        %% Sink

        subgraph Sink
            direction LR

            v1("<b>[Line: 27]</b> sink(d2)")
        end
    end
    %% Class Assignment
    Source:::invis
    Sink:::invis

    Traces0:::invis
    File0:::invis

    %% Connections

    Source --> Traces0
    Traces0 --> Sink

    %% Clickable

    click v0 href "https://github.semgrep.dev/repo/blob/12345/foo.py#L26" "View in source" _blank
    click v1 href "https://github.semgrep.dev/repo/blob/12345/foo.py#L27" "View in source" _blank
    click v2 href "https://github.semgrep.dev/repo/blob/12345/foo.py#L26" "View in source" _blank
```

</details>

###### 🔴 Fix or [ignore](https://semgrep.dev/docs/ignoring-files-folders-code/#reference-summary) this finding to merge your pull request. [Ignore this finding](https://semgrep.dev/docs/ignoring-files-folders-code) from [dataflow-example](https://semgrep.dev/r/python.lang.correctness.useless-eqeq.useless-eqeq?utm_campaign=finding_notification&utm_medium=review_comment&utm_source=bitbucket&utm_content=rule).
<!--

🤫 WELCOME TO SECRET SEMGREP! 🤫
This information is for debugging purposes and does not appear in rendered PR comments.

Finding id: 27428
Syntactic id: XQOytkHsywzqyAOHLIIa
Start line: 3,1
End line: 3,7
Index: 0
-->